module/bdev/ocf/vbdev_ocf.c

```diff
@@ -1,1794 +1,2833 @@
-		return;
+		SPDK_ERRLOG("OCF core '%s': failed to unregister OCF vbdev during cache stop: %s\n",
+			    ocf_core_get_name(core), spdk_strerror(-error));
 	}
 
-	ocf_mngt_cache_stop(vbdev->ocf_cache, stop_vbdev_cmpl, vbdev);
+	ocf_mngt_cache_put(cache);
+	vbdev_ocf_core_destroy(core_ctx);
+
+	if (ocf_cache_get_core_count(cache) == ocf_cache_get_core_inactive_count(cache)) {
+		/* All cores in this cache were already unregistered
+		 * and detached, so proceed with stopping the cache. */
+		ocf_mngt_cache_lock(cache, _cache_stop_lock_cb, mngt_ctx);
+	}
 }
 
-/* Stop OCF cache object
- * vbdev_ocf is not operational after this */
-static void
-stop_vbdev(struct vbdev_ocf *vbdev)
+static int
+_cache_stop_core_visitor(ocf_core_t core, void *cb_arg)
 {
-	if (!is_ocf_cache_running(vbdev)) {
-		vbdev_ocf_mngt_continue(vbdev, 0);
-		return;
+	struct vbdev_ocf_mngt_ctx *mngt_ctx = cb_arg;
+	ocf_cache_t cache = ocf_core_get_cache(core);
+	struct vbdev_ocf_core *core_ctx = ocf_core_get_priv(core);
+	int rc = 0;
+
+	SPDK_DEBUGLOG(vbdev_ocf, "OCF core '%s': cache stop visit\n", ocf_core_get_name(core));
+
+	if (!core_ctx) {
+		/* Skip this core. If there is no context, it means that this core
+		 * was added from metadata during cache load and it's just an empty shell. */
+		ocf_mngt_cache_put(cache);
+		return 0;
 	}
 
-	if (!g_fini_started && get_other_cache_instance(vbdev)) {
-		SPDK_NOTICELOG("Not stopping cache instance '%s'"
-			       " because it is referenced by other OCF bdev\n",
-			       vbdev->cache.name);
-		vbdev_ocf_mngt_continue(vbdev, 0);
-		return;
+	/* If core is detached it's already unregistered, so just free its data and exit. */
+	if (!vbdev_ocf_core_is_base_attached(core_ctx)) {
+		ocf_mngt_cache_put(cache);
+		vbdev_ocf_core_destroy(core_ctx);
+		return 0;
 	}
 
-	ocf_mngt_cache_lock(vbdev->ocf_cache, stop_vbdev_cache_lock_cmpl, vbdev);
-}
+	core_ctx->mngt_ctx = mngt_ctx;
 
-static void
-flush_vbdev_cmpl(ocf_cache_t cache, void *priv, int error)
-{
-	struct vbdev_ocf *vbdev = priv;
+	if ((rc = vbdev_ocf_core_unregister(core_ctx, _cache_stop_core_unregister_cb, core))) {
+		SPDK_ERRLOG("OCF core '%s': failed to start unregistering OCF vbdev: %s\n",
+			    ocf_core_get_name(core), spdk_strerror(-rc));
+		ocf_mngt_cache_put(cache);
+		return rc;
+	}
 
-	ocf_mngt_cache_unlock(cache);
-	vbdev_ocf_mngt_continue(vbdev, error);
+	return rc;
 }
 
-static void
-flush_vbdev_cache_lock_cmpl(ocf_cache_t cache, void *priv, int error)
+static int
+_module_fini_cache_visitor(ocf_cache_t cache, void *cb_arg)
 {
-	struct vbdev_ocf *vbdev = (struct vbdev_ocf *)priv;
+	int i, rc = 0;
 
-	if (error) {
-		SPDK_ERRLOG("Error %d, can not lock cache instance %s\n",
-			    error, vbdev->name);
-		vbdev_ocf_mngt_continue(vbdev, error);
-		return;
+	SPDK_DEBUGLOG(vbdev_ocf, "OCF cache '%s': module stop visit\n", ocf_cache_get_name(cache));
+
+	/* Increment cache refcount to not destroy cache structs before destroying all cores. */
```

Comment:
This sounds like a workaround. I think we should have a proper management of cache lifetime.